### PR TITLE
feat(perl): Added perl version module

### PR DESCRIPTION
**Pull request by stefb69**
#### Description
I have added a module for perl support, based on the php or ruby module.

#### Motivation and Context

I use some perl and tought it should be suppoted in 

#### Screenshots (if appropriate):
image

#### How Has This Been Tested?

- [ ] I have tested using **MacOS**
- [x] I have tested using **Linux**
- [ ] I have tested using **Windows/WSL2**
I don't think there could be much regression. Tested on wsl2 with zsh. linux

#### Checklist:

- [x] I have updated the documentation accordingly.
- [x] I have updated the tests accordingly.
But i've done only the english doc.

### CODE REVIEW

File: docs/fr-FR/config/README.md [hunk 1]
```diff
@@ -1100,6 +1101,36 @@ The `ocaml` module shows the currently installed version of OCaml. Le module est
 symbol = "🐪 "
 ```
 
+## Perl
```
Comment (ATiltedTree) [inline docs/fr-FR/config/README.md:1104]: The translations are automatically updated by crowdin. So please revert all the changes to the lang files.

File: src/modules/perl.rs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+            ".perl-version",
+        ])
+        .set_extensions(&["pl", "pm"])
+        .is_match();
+
+    if !is_perl_project {
+        return None;
+    }
+
+    let perl_version = utils::exec_cmd("perl", &["-e", "print $^V;"])?.stdout;
```
Comment (ATiltedTree) [inline src/modules/perl.rs:28]: Your command does not match with your mock command. So change either one to match.

File: src/modules/perl.rs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    #[test]
+    fn folder_with_cpanfile_file() -> io::Result<()> {
+        let dir = tempfile::tempdir()?;
+        File::create(dir.path().join("cpanfile"))?.sync_all()?;
+
+        let actual = render_module("perl", dir.path(), None);
+
+        let expected = Some(format!(
+            "via {} ",
+            Color::Fixed(149).bold().paint("🐪 v5.26.1")
```
Comment (ATiltedTree) [inline src/modules/perl.rs:67]: The version needs to match with the output of the mock command.

Comment (stefb69) [reply src/modules/perl.rs:67]: yes but test env all have different versions of perl. Linux 5.26.1, macos 5.30.2 and windows 5.30.3

Comment (ATiltedTree) [reply src/modules/perl.rs:67]: The problem is that the command you are running when getting the version in the module is different from the command in the mock function. Which means that the command is run directly and the output of the mock command is not used.

Comment (stefb69) [reply src/modules/perl.rs:67]: ok I think i fixed that in last commit

Comment (ATiltedTree) [reply src/modules/perl.rs:67]: Checks are still failing. You need to replace the mock command with `perl -e print $^V;`.

File: src/utils.rs [hunk 4]
```diff
@@ -87,6 +87,10 @@ active boot switches: -d:release\n",
             stdout: String::from("4.10.0"),
             stderr: String::default(),
         }),
+        "perl -e 'print $^V'" => Some(CommandOutput {
```
Comment (ATiltedTree) [inline src/utils.rs:90]:
```suggestion
        "perl -e print $^V;" => Some(CommandOutput {
```

File: docs/config/README.md [hunk 5]
```diff
@@ -1179,6 +1180,37 @@ The module will be shown if any of the following conditions are met:
 symbol = "🐪 "
 ```
 
+## Perl
+
+The `perl` module shows the currently installed version of Perl.
+The module will be shown if any of the following conditions are met:
+
+- The current directory contains a `Makefile.PL` file
```
Comment (vladimyr) [inline docs/config/README.md:1188]: There is also `Build.PL` used by [`Module::Build`](https://metacpan.org/pod/Module::Build) and [`Module::Build::Tiny`](https://metacpan.org/pod/Module::Build::Tiny).

File: docs/config/README.md [hunk 6]
```diff
@@ -1179,6 +1180,37 @@ The module will be shown if any of the following conditions are met:
...
+- The current directory contains a `cpanfile` file
```
Comment (vladimyr) [inline docs/config/README.md:1189]: Well, if you support `cpanfile` then you should probably support  `cpanfile.snapshot` too. That's pretty much CPAN's version of a lockfile.

File: docs/config/README.md [hunk 7]
```diff
@@ -1179,6 +1180,37 @@ The module will be shown if any of the following conditions are met:
...
+- The current directory contains a `META.json` file
+- The current directory contains a `META.yml` file
+- The current directory contains a `.perl-version` file
+- The current directory contains a `.pl` file
+- The current directory contains a `.pm` file
```
Comment (vladimyr) [inline docs/config/README.md:1194]: There are also documentation files aka `.pod` files.

File: docs/config/README.md [hunk 8]
```diff
@@ -1179,6 +1180,37 @@ The module will be shown if any of the following conditions are met:
...
+| `style`    | `"bold 149"` | The style for the module.                             |
+| `disabled` | `false`      | Disables the `perl` module.                           |
+
+### Example
+
+```toml
+# ~/.config/starship.toml
+
+[perl]
+symbol = "🐫 "
```
Comment (vladimyr) [inline docs/config/README.md:1210]: Nasty, I'm sure OCaml folks would loudly object to the abuse of their camel 😀

File: src/utils.rs [hunk 9]
```diff
@@ -87,6 +87,10 @@ active boot switches: -d:release\n",
             stdout: String::from("4.10.0"),
             stderr: String::default(),
         }),
+        "perl -e print $^V;" => Some(CommandOutput {
```
Comment (vladimyr) [inline src/utils.rs:90]: This won't work cross-platform, in fact, it didn't on my Android. To make it work use the following form: `perl -e 'printf "%vd", $^V'`

Comment (stefb69) [reply src/utils.rs]: ok i've used  perl -e 'printf q#%vd#,$^V;' to try to avoid any double escape condition

File: docs/config/README.md (hunk 8 above)
Comment (stefb69) [reply docs/config/README.md]: this was a joke, as just the same thing is done a few lines up, i replace example by an oyster

File: docs/config/README.md (hunk 7 above)
Comment (stefb69) [reply docs/config/README.md:1194]: added .pod file

Comment (stefb69) [reply docs/config/README.md:1196]: added, doc updated

File: docs/config/README.md (hunk 8 above)
Comment (vladimyr) [reply docs/config/README.md]: My comment was also a joke 🤣

File: docs/config/README.md (hunk 6 above)
Comment (stefb69) [reply docs/config/README.md:1190]: added, doc updated, though i think there has to be a cpanfile if there is a cpanfile.snapshot, not sure, it wont hurt anyhow.

File: src/utils.rs (hunk 9 above)
Comment (vladimyr) [reply src/utils.rs]: Perfect 👍

File: docs/config/README.md (hunk 5 above)
Comment (stefb69) [reply docs/config/README.md:1188]: added, doc updated

File: docs/config/README.md (hunk 6 above)
Comment (vladimyr) [reply docs/config/README.md:1190]: > added, doc updated, though i think there has to be a cpanfile if there is a cpanfile.snapshot, not sure, it wont hurt anyhow.

That would be pretty unusual, I agree... But as you said, won't hurt 🙃

File: src/modules/perl.rs [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+            ".perl-version",
+        ])
+        .set_extensions(&["pl", "pm", "pod"])
+        .is_match();
+
+    if !is_perl_project {
+        return None;
+    }
+
+    let perl_version = utils::exec_cmd("perl", &["-e", "printf q#%vd#,$^V;"])?.stdout;
```
Comment (vladimyr) [inline src/modules/perl.rs:30]:
```suggestion
    let perl_version = utils::exec_cmd("perl", &["-e", "'printf q#%vd#,$^V;'"])?.stdout;
```
Note the inner single quotes, without them tests won't get mocked but real Perl version.

Comment (stefb69) [reply src/modules/perl.rs]: ok done

Comment (stefb69) [reply src/modules/perl.rs]: thanks for you help this is my first PR in rust

Comment (stefb69) [reply src/modules/perl.rs]: And now all check pass, not just ubuntu, cool.

Comment (vladimyr) [reply src/modules/perl.rs]: Let me tell you a _secret_, I mostly do node 🤣

Comment (stefb69) [reply src/modules/perl.rs]: There is a regression with this modification
image

Comment (stefb69) [reply src/modules/perl.rs:30]: resolved by 916e1b1

File: src/utils.rs [hunk 11]
```diff
@@ -101,6 +101,10 @@ active boot switches: -d:release\n",
             stdout: String::from("4.08.1\n"),
             stderr: String::default(),
         }),
+        "perl -e printf q#%vd#,$^V;" => Some(CommandOutput {
```
Comment (vladimyr) [inline src/utils.rs:104]:
```suggestion
        "perl -e 'printf q#%vd#,$^V;'" => Some(CommandOutput {
```

File: src/modules/perl.rs (hunk 10 above)
Comment (vladimyr) [reply src/modules/perl.rs:31]: Actually I was right back then, this should be quoted. The reason it failed is that mock also needed adjustments.

Comment (stefb69) [reply src/modules/perl.rs:31]: nope. mock has not failed it's some other module. with quotes it fails

File: docs/config/README.md [hunk 12]
```diff
@@ -1661,6 +1662,40 @@ The module will be shown if any of the following conditions are met:
...
+
+- The current directory contains a `Makefile.PL` file
+- The current directory contains a `Build.PL` file
+- The current directory contains a `cpanfile` file
+- The current directory contains a `cpanfile.snapshot` file
+- The current directory contains a `META.json` file
+- The current directory contains a `META.yml` file
+- The current directory contains a `.perl-version` file
+- The current directory contains a `.pl` file
+- The current directory contains a `.pm` file
```
Comment (vladimyr) [inline docs/config/README.md:1678]:
```suggestion
- The current directory contains a `Makefile.PL` or `Build.PL` file
- The current directory contains a `cpanfile` or `cpanfile.snapshot` file
- The current directory contains a `META.json` or `META.yml` file
- The current directory contains a `.perl-version` file
- The current directory contains a file with `.pl`, `.pm` or `.pod` extension
```

Comment (vladimyr) [reply docs/config/README.md:1678]: It is mostly a cosmetic suggestion for sake of decreasing the number of lines.

File: docs/config/README.md [hunk 13]
```diff
@@ -1661,6 +1662,40 @@ The module will be shown if any of the following conditions are met:
...
+| `disabled` | `false`      | Disables the `perl` module.                           |
+
+### Example
+
+```toml
+# ~/.config/starship.toml
+
+[perl]
+symbol = "🦪 "
+```
```
Comment (vladimyr) [inline docs/config/README.md:1696]: ### Options

| Option     | Default                            | Description                                             |
| ---------- | ---------------------------------- | ------------------------------------------------------- |
| `format`   | `"via [$symbol$version]($style) "` | The format string for the module.                       |
| `symbol`   | `"🐪 "`                            | The symbol used before displaying the version of Perl. |
| `style`    | `"bold 149"`                    | The style for the module.                               |
| `disabled` | `false`                            | Disables the `perl` module.                            |

### Variables

| Variable | Example   | Description                          |
| -------- | --------- | ------------------------------------ |
| version  | `v5.26.1` | The version of `perl`               |
| symbol   |           | Mirrors the value of option `symbol` |
| style\*  |           | Mirrors the value of option `style`  |

\*: This variable can only be used as a part of a style string

### Example

```toml
# ~/.config/starship.toml

[perl]
format = "via [🦪 $version]($style) "
```

File: docs/config/README.md [hunk 14]
```diff
@@ -1661,6 +1662,40 @@ The module will be shown if any of the following conditions are met:
...
+- The current directory contains a `.perl-version` file
+- The current directory contains a `.pl` file
+- The current directory contains a `.pm` file
+- The current directory contains a `.pod` file
+
+### Options
+
+| Variable   | Default      | Description                                           |
+| ---------- | ------------ | ----------------------------------------------------- |
+| `symbol`   | `"🐪 "`      | The symbol used before displaying the version of PHP. |
```
Comment (vladimyr) [inline docs/config/README.md:1685]: You have PHP leftover here.

File: src/modules/perl.rs [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+
+use crate::configs::perl::PerlConfig;
+use crate::formatter::StringFormatter;
+use crate::utils;
+
+/// Creates a module with the current perl version
+///
+/// Will display the perl version if any of the following criteria are met:
+///     - Current directory contains a `.pl` or a `.pm` file
+///     - Current directory contains a "Makefile.PL", "cpanfile", "META.json", "META.yml", or ".perl-version" file
```
Comment (vladimyr) [inline src/modules/perl.rs:11]: Please update this comment section. I believe you are missing files with `.pod` extension, `Build.PL` file and `cpanfile.snapshot`.

File: src/utils.rs (hunk 11 above)
Comment (stefb69) [reply src/utils.rs:104]: With your modifcations tests fails..., just tested on my machine.

image

File: docs/config/README.md (hunk 14 above)
Comment (stefb69) [reply docs/config/README.md:1685]: oups

File: src/utils.rs (hunk 11 above)
Comment (vladimyr) [reply src/utils.rs:104]: Let's see. I started with:
```
starship on  perl-module [$!] is 📦 v0.44.0 via 🦀 v1.44.1
❯ git show --shortstat
commit 481401be552a5c42552e006a7b9bb89dc38e7603 (HEAD -> perl-module)
Author: Stephane Benoit <stefb@wizzz.net>
Date:   Mon Jul 27 20:50:39 2020 +0200

    Fixed missing new line

 1 file changed, 1 insertion(+), 1 deletion(-)
```

Then, I made changes both to mock and actual invocation:
```diff
diff --git a/src/modules/perl.rs b/src/modules/perl.rs
index 08fd824..1854ac2 100644
--- a/src/modules/perl.rs
+++ b/src/modules/perl.rs
@@ -28,7 +28,7 @@ pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
         return None;
     }
 
-    let perl_version = utils::exec_cmd("perl", &["-e", "printf q#%vd#,$^V;"])?.stdout;
+    let perl_version = utils::exec_cmd("perl", &["-e", "'printf q#%vd#,$^V;'"])?.stdout;
 
     let mut module = context.new_module("perl");
     let config: PerlConfig = PerlConfig::try_load(module.config);
diff --git a/src/utils.rs b/src/utils.rs
index c83bea6..373fec2 100644
--- a/src/utils.rs
+++ b/src/utils.rs
@@ -101,7 +101,7 @@ active boot switches: -d:release\n",
             stdout: String::from("4.08.1\n"),
             stderr: String::default(),
         }),
-        "perl -e printf q#%vd#,$^V;" => Some(CommandOutput {
+        "perl -e 'printf q#%vd#,$^V;'" => Some(CommandOutput {
             stdout: String::from("5.26.1"),
             stderr: String::default(),
         }),

```

For the record, I'm having Perl 5.32.0 installed locally:
```
starship on  perl-module [$!] is 📦 v0.44.0 via 🦀 v1.44.1
❯ perl --version

This is perl 5, version 32, subversion 0 (v5.32.0) built for darwin-thread-multi-2level

Copyright 1987-2020, Larry Wall

Perl may be copied only under the terms of either the Artistic License or the
GNU General Public License, which may be found in the Perl 5 source kit.

Complete documentation for Perl, including FAQ lists, should be found on
this system using "man perl" or "perldoc perl".  If you have access to the
Internet, point your browser at http://www.perl.org/, the Perl Home Page.
```

I've run the test:
```
starship on  perl-module [$!] is 📦 v0.44.0 via 🦀 v1.44.1 took 1m6s
❯ cargo test modules::perl::folder_with_perl_files
warning: method is never used: `set_value`
  --> src/segment.rs:46:5
   |
46 | /     pub fn set_value<T>(&mut self, value: T) -> &mut Self
47 | |     where
48 | |         T: Into<String>,
49 | |     {
50 | |         self.value = value.into();
51 | |         self
52 | |     }
   | |_____^
   |
   = note: `#[warn(dead_code)]` on by default

warning: 1 warning emitted

warning: method is never used: `is_empty`
   --> src/module.rs:101:5
    |
101 |     pub fn is_empty(&self) -> bool {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `#[warn(dead_code)]` on by default

warning: method is never used: `new`
  --> src/segment.rs:21:5
   |
21 |     pub fn new(name: &str) -> Self {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: method is never used: `set_value`
  --> src/segment.rs:46:5
   |
46 | /     pub fn set_value<T>(&mut self, value: T) -> &mut Self
47 | |     where
48 | |         T: Into<String>,
49 | |     {
50 | |         self.value = value.into();
51 | |         self
52 | |     }
   | |_____^

warning: method is never used: `is_empty`
  --> src/segment.rs:68:5
   |
68 |     pub fn is_empty(&self) -> bool {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: 4 warnings emitted

    Finished test [unoptimized + debuginfo] target(s) in 0.74s
     Running target/debug/deps/starship-2b44a645abddee21

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 251 filtered out

     Running target/debug/deps/starship-b3862c2e5a1d1776

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 263 filtered out

     Running target/debug/deps/testsuite-d9f5f469e468477c

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 181 filtered out


```

And it passed ✅

Comment (vladimyr) [reply src/utils.rs:104]: @stefb69 It seems that the key to success is to quote on both sides (invocation and mock). Let me know if that works for you 🤞

File: docs/config/README.md (hunk 14 above)
Comment (vladimyr) [reply docs/config/README.md]: Mah, it happens all the time 😃

File: docs/config/README.md [hunk 16]
```diff
@@ -1661,6 +1662,45 @@ The module will be shown if any of the following conditions are met:
...
+- The current directory contains a `cpanfile` or `cpanfile.snapshot` file
+- The current directory contains a `META.json` file or `META.yml` file
+- The current directory contains a `.perl-version` file
+- The current directory contains a `.pl`, `.pm` or `.pod`
+
+### Options
+
+| Variable   | Default                            | Description                                           |
+| ---------- |----------------------------------- | ----------------------------------------------------- |
+| `format`   | `"via [$symbol$version]($style) "` | The symbol used before displaying the version of Perl |
```
Comment (vladimyr) [inline docs/config/README.md:1681]: I'm sorry 🙃 
```suggestion
| `format`   | `"via [$symbol$version]($style) "` | The format string for the module.                       |
```

File: src/utils.rs (hunk 11 above)
Comment (stefb69) [reply src/utils.rs:104]: I see your point but please try the resulting binary in a perl folder : 
image
Same as last time, i already tried and had the same issue. The only way everything works ok is without quotes

Comment (stefb69) [reply src/utils.rs:104]: And actually the tests do not fail in the automated tests, it is another module that fails with ubuntu nightly and macos nightly

Comment (stefb69) [reply src/utils.rs:104]: image

Comment (vladimyr) [reply src/utils.rs:104]: > And actually the tests do not fail in the automated tests, it is another module that fails with ubuntu nightly and macos nightly

You can ignore those, they fail sporadically on other PRs too...

Comment (vladimyr) [reply src/utils.rs:104]: > I see your point but please try the resulting binary in a perl folder :
> image
> Same as last time, i already tried and had the same issue. The only way everything works ok is without quotes

Wow, you are right. This is interesting 🤔

Comment (vladimyr) [reply src/utils.rs:104]: Well, what can I say? I was wrong about this and sorry for pushing you into reiterating on what should've stayed closed/resolved topic 🤦 
Actually this provided me with a valuable lesson 👍  and now I need to go and revert portions of #1491 (like this for instance https://github.com/starship/starship/pull/1491/files#diff-3dbd16223f25e8b5e0bcf099743107e9R104)

**Comment by andytom** [comment]
Hi @stefb69, I think we have fixed the CI issues on OSX if you can give this a rebase it should be ready to merge.

**Comment by stefb69** [comment]
> 
> 
> Hi @stefb69, I think we have fixed the CI issues on OSX if you can give this a rebase it should be ready to merge.

Hi @andytom, I've just synced this mr with upstream. 
The build tests are ok.

**Pull request merged**